crates/agglayer-prover/src/executor.rs

```diff
@@ -0,0 +1,275 @@
+            async { Err(Error::NetworkProverDisabled) }.boxed()
+        };
+
+        let mut local = self.local.clone();
+
+        let fut = async move {
+            match network.await {
+                Ok(res) => Ok(res),
+                Err(err) => {
+                    error!("Network prover failed: {:?}", err);
+
+                    local.ready().await?.call(req).await
+                }
+            }
+        };
+
+        Box::pin(fut)
+    }
+}
+
+#[derive(Clone)]
+struct LocalExecutor {
+    proving_key: SP1ProvingKey,
+    verification_key: SP1VerifyingKey,
+    prover: Arc<CpuProver>,
+    timeout: Duration,
+}
+
+impl Service<Request> for LocalExecutor {
+    type Response = Response;
+
+    type Error = Error;
+
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, _req: Request) -> Self::Future {
+        let prover = self.prover.clone();
+        let stdin = SP1Stdin::new();
+        let opts = ProofOpts {
+            timeout: Some(self.timeout),
+            ..Default::default()
+        };
+        let kind = SP1ProofKind::default();
+
+        let proving_key = self.proving_key.clone();
+        let verification_key = self.verification_key.clone();
+
+        Box::pin(
+            spawn_blocking(move || {
+                let context = SP1Context::default();
+                let proof = prover
+                    .prove(&proving_key, stdin, opts, context, kind)
+                    .map_err(|error| Error::ProverFailed(error.to_string()))?;
+
+                prover
+                    .verify(&proof, &verification_key)
+                    .map_err(|error| Error::ProofVerificationFailed(error.into()))?;
+
+                Ok(Response { proof })
+            })
+            .map_err(|_| Error::UnableToExecuteProver)
+            .and_then(|res| async { res }),
+        )
+    }
+}
+
+#[derive(Clone)]
+struct NetworkExecutor {
+    prover: Arc<NetworkProver>,
+    verification_key: SP1VerifyingKey,
+    timeout: Duration,
+}
+
+impl Service<Request> for NetworkExecutor {
+    type Response = Response;
+
+    type Error = Error;
+
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        let prover = self.prover.clone();
+        let stdin = req.into();
+
+        let verification_key = self.verification_key.clone();
+        let timeout = self.timeout;
+
+        let fut = async move {
+            let proof = prover
+                .prove(
+                    ELF,
+                    stdin,
+                    sp1_sdk::proto::network::ProofMode::default(),
+                    Some(timeout),
```

Comment:
This has, I believe, the same issue as the Cpu prover timeout mentioned before. The tower layer timeout will probably terminate first.